### DISCUSSION TITLE
/create-authentication-provider-visual-studio-code/

### DISCUSSION THREAD

**Discussion by comment-migrator** [estruyf/eliostruyf-comments#596] [Comments]
Create an Authentication Provider for Visual Studio Code
        
[https://www.eliostruyf.com/create-authentication-provider-visual-studio-code/](https://www.eliostruyf.com/create-authentication-provider-visual-studio-code/)

**Comment by comment-migrator** [comment]
```header
Originally posted by Rahul Vr on 2022-06-14 08:19
```

sorry, works fine. thanks a lot

**Reply by comment-migrator** [reply]
```header
Originally posted on 2022-06-14 13:59
```

No problem, glad to hear you found it.

**Comment by comment-migrator** [comment]
```header
Originally posted by Maochen Luo on 2022-07-26 11:54
```

Is there any document about google login AuthenticationProvider，Is it the same as this? Thank you

**Reply by comment-migrator** [reply]
```header
Originally posted on 2022-07-26 15:29
```

OAuth for Google works similar, so I expect you'll need to implement the correct endpoints and flows.

**Comment by milidubey** [comment]
Thank you for this article, it was very helpful!

**Comment by npreiser** [comment]
I implemented your code and it works great from within vscode.   I am trying to use it now from inside Theia IDE.   The plugin is partially working however there is some issues around the redirect url.   After login, when the browser tries to send back the token,  basically the urieventhandler does not fire.  When running in Linux we see the entire things is comming back from Auth0,  not sure why the urihandler is not picking it up.  Are there any known restrictions with the naming of the redicrect handler..etc?...  My urischeme is  theia://...x..x.xxx,  not vscode://   ???,  thanks .

**Reply by estruyf** [reply]
Hi @npreiser, I'm not aware of any restrictions. 

1. When you use `theia://`, does it trigger it to open the IDE?
2. Does the IDE ask you to open the link within your extension? This happens at least the first time in VS Code.

Also verify if your extension name is correct: `<publisher>.<extension-name>`

**Comment by nmcostello** [comment]
I implemented this using oauth instead of specifically auth0. Obviously this is pretty similar there are just a few little differences based on my company's implementation. 

I added some refresh token logic, but I don't think I have enough understanding to implement the second point:
> Validate if your refresh token is still valid. If not, remove the session.

Where would that be? I don't necessarily know how to interact with the auth provider besides during the `getSession` command during the extension activation. My naive approach would be to set `forceNewSession: true` and then check for expiry when `AuthenticationProvider.createSession` is called. 

Any feedback would be appreciated.

**Reply by nmcostello** [reply]
I ended up writing something like this (it's not completely fleshed out and tested yet): 
```TypeScript
const createOAuthSession = async (authenticationProvider: OAuthAuthenticationProvider) => {
  let session = await vscode.authentication.getSession("oauth", [], { createIfNone: true }) as OAuthAuthenticationSession;
  if (!session) {
    vscode.window.showInformationMessage(`No auth session was started`);
  }

  // check to see if the refreshToken has expired
  if (!session.refreshTokenExpiresAt || session.refreshTokenExpiresAt < Date.now()) {
    console.log("Refresh token has expired. Restarting login process");
    authenticationProvider.removeSession(session.id);
    session = await vscode.authentication.getSession("oauth", [], { forceNewSession: true }) as OAuthAuthenticationSession;
  }

  // check to see if the session is expired and needs to be refreshed
  if (session.accessToken && (!session.expiresAt || session.expiresAt < Date.now())) {
		vscode.window.showInformationMessage(`Welcome back ${session.account.label}`);
    // trigger a refresh
    try {
      authenticationProvider.triggerRefresh(session);
    } catch (e) {
      vscode.window.showInformationMessage(`Failed to refresh token. Restarting login process.`);
      authenticationProvider.removeSession(session.id);
      session = await vscode.authentication.getSession("oauth", [], { forceNewSession: true }) as OAuthAuthenticationSession;
    }
  ```

**Reply by estruyf** [reply]
Hi @nmcostello, your logic is very similar to what I once implemented, but I've put the logic in the `getSessions` method. That way, when you call: `let session = await vscode.authentication.getSession("oauth", [], { createIfNone: false })`, you can validate the token whenever you want to start using it in your extension.

```ts
public async getSessions(): Promise<readonly Auth0AuthenticationSession[]> {
  try {
    const allSessions = await this.context.secrets.get(SESSIONS_SECRET_KEY)
    if (allSessions) {
      const sessions = JSON.parse(allSessions);

      // Refresh the session
      if (sessions.length > 0) {
        const session = sessions[0];
        const refreshedSession = await this.refreshSession(session);

        if (refreshedSession) {
          return [refreshedSession];
        }
      }

      return [];
    }
  } catch (err: unknown) {
    await this.context.secrets.store(SESSIONS_SECRET_KEY, JSON.stringify([]));
  }
  return [];
}
```

In the `refreshSession` method, I have implemented the logic to verify if the access token is still valid, or that it needs to be refreshed.

**Reply by nmcostello** [reply]
Wonderful! I would not have figured this out without your blog. I owe you a coffee. Thank you :)

**Reply by estruyf** [reply]
Thank you, @nmcostello, for the coffee.

**Comment by nmcostello** [comment]
In your example video, you were able to trigger the login via the Accounts options. I haven't been able to replicate that behavior and can only login on startup or refresh. Any idea how I can get that option to appear?

**Reply by estruyf** [reply]
@nmcostello in case you want to show a numbered badge on the accounts activity bar icon, you need to make use of the `createIfNone` property when retrieving an authenticated session.

```ts
const session = await vscode.authentication.getSession("auth0", [], { createIfNone: false });
```

From the VS Code types docs:

```ts
/**
 * Whether login should be performed if there is no matching session.
 *
 * If true, a modal dialog will be shown asking the user to sign in. If false, a numbered badge will be shown
 * on the accounts activity bar icon. An entry for the extension will be added under the menu to sign in. This
 * allows quietly prompting the user to sign in.
 *
 * If there is a matching session but the extension has not been granted access to it, setting this to true
 * will also result in an immediate modal dialog, and false will add a numbered badge to the accounts icon.
 *
 * Defaults to false.
 *
 * Note: you cannot use this option with {@link AuthenticationGetSessionOptions.silent silent}.
 */
createIfNone?: boolean;
```

**Comment by AnnaSalata** [comment]
This article helped me a lot with creating an auth provider in my extension, thanks! If you have any information about tracking of checking/unchecking the extension as trusted under _"Manage Trusted Accounts"_ menu option of the extension's account, could you please help? I'm experiencing the following issue: while unchecking, _"Grant access to {Auth Provider} for {Extension}... (1)"_ menu item appears under Accounts and disappears when I click it, but if I just go to _"Manage Trusted Accounts"_ and check my extension again, _"Grant access to {Auth Provider} for {Extension}... (1)"_  stays under Accounts. Should'nt both ways of access granting work the same way?

**Reply by estruyf** [reply]
Hi @AnnaSalata could you share some screenshots? That would make it easier to follow.

The `Manage Trusted Accounts` for the Authentication Provider is something VS Code manages, so you have less impact on this, or you need to do a check if you can retrieve a session.

**Reply by AnnaSalata** [reply]
Hi @estruyf! I have a full description in the bug I've opened on github, just wasn't sure, if posting links is allowed here. Could you check it, please?
https://github.com/microsoft/vscode/issues/227694

_or you need to do a check if you can retrieve a session._
Yes, I can retrieve a session, but when "Grant access to {Auth Provider} for {Extension}... (1)" stays under Accounts, user can later sign in, using another username, then press Grant access and two account for one extension appears even if auth provider is configured with { supportsMultipleAccounts: false }

**Comment by nmcostello** [comment]
Hi, I wanted to start by thanking you for all the help you've provided in the past. Your contributions have been really valuable to me.

I've recently encountered a limitation with the current implementation, specifically when using multiple VSCode windows. When logging in and refreshing sessions, I've noticed some bugs that can cause issues. It seems that one window isn't aware of changes made in other windows, leading to situations where all open instances of VSCode refresh the session or relogin simultaneously. From a UI/UX perspective, this isn't ideal, as it can be confusing and disrupt the user experience.

I was wondering if you've come across this issue before or have any recommendations for resolving it. Your insight and expertise would be really helpful in addressing this problem.

**Reply by estruyf** [reply]
I haven't encountered it, but I can see it happening. In such a case, it would be best to have some logic where you add a status to the VSCode global state, which each of your instances can read. If one starts the login process, the others wait until it is completed.

**Comment by samir-br-ba** [comment]
Hi Elio, I have quick question.
I have implemented something similar with redirection link for auth (custom backend) and we have visual bug where Sign in button in Settings/Account context menu is appearing twice, when I log in I still see another button saying Sign in to....
We have debugged everything, provider is registering only once, it is called only once, everything looks so normal, but still we are seeing duplicate Sign in button. Do you have any idea what could be the issue or what should I try/look at etc..

Appreciate your help, thank you!

**Reply by samir-br-ba** [reply]
Heeeelp :D

**Reply by estruyf** [reply]
Can you share a screenshot please. If you reload VSCode, does it still appear twice? Or is it only the first time?

**Reply by samir-br-ba** [reply]
Checkout the link, it remains the same once I restart extension

https://imagekit.io/tools/asset-public-link?detail=%7B%22name%22%3A%22auth.png%22%2C%22type%22%3A%22image%2Fpng%22%2C%22signedurl_expire%22%3A%222028-03-26T09%3A14%3A00.867Z%22%2C%22signedUrl%22%3A%22https%3A%2F%2Fmedia-hosting.imagekit.io%2F6fda966dafe64706%2Fauth.png%3FExpires%3D1837674841%26Key-Pair-Id%3DK2ZIVPTIP2VGHC%26Signature%3DEz6O7IUQ52~pb2B5yLSIN0Dd51AhHn22Eo2cR6MwHGHBAgpRc2Y25X6HCo9v7KhNe9qpEwtIiMTNLHFmZaYNxeMJjDMQPmheKfNAcqyEDlyfuh9o9cRLJEioV1ZU6zGdwmA05nr-F9vPp-gMuHFyC4gX7uIjF9j5qu6F~ypsn947SJQvOSVZDFjItLqtyRt1FotKjk~ZlhIf1fb07gw~SY-za2FR-PwUi8hAGHMiFeCQgW18xAE~I2P~jYu7ZxuxonkNNpfteqr-JxwkJoxjT5-fnzRRL7hTrbQkmy172iBcR3fuC2MhEKkHwBvYG7AFGPU9HQutRQRXwhl8ZucOzw__%22%7D

**Reply by estruyf** [reply]
That's indeed weird. I can't tell where it would go wrong without a code review.

**Comment by regosen** [comment]
Hi, how would you implement promiseFromEvent?

**Reply by regosen** [reply]
Ah, never mind, I found your sample code!  https://github.com/estruyf/vscode-auth-sample/blob/main/src/util.ts
